### ISSUE TITLE
AWS is returning previous user session.

### ISSUE THREAD

**Issue by shahan-ayyub** [aws-amplify/amplify-ios#1570]
### Describe the bug

This issue is in continuation with our reported bug here:
https://github.com/aws-amplify/amplify-ios/issues/1360

As per the fix for 1360 it seems that our issue for username has been resolved. However, now another issue is observed where previous user who was signed out is successfully logged in.

Pre-requisites to run POC:
1- pod install
2- replace '*' with necessary ids/info in amplifyconfiguration.json
3- provide two different users' data in ViewController.swift (Line# 14)

POC to reproduce issue:
[AmplifyPOC.zip](https://github.com/aws-amplify/amplify-ios/files/7726691/AmplifyPOC.zip)

### Steps To Reproduce

```swift
Steps within POC to reproduce issue:

1. Tap "Signin User1" button.
2. Wait for success
3. Tap "Signin User2" button.
4. While log in is in process from step 3, Tap "Signin User2" again in POC.
5. See console logs in POC. You will notice that first "User 1" will be returned then "User 2" will be returned.

NOTE: In the POC, whenever a user is being logged in, the login call is preceded by log out.
```

### Expected behavior

It should fetch the correct/new user session rather than old user session.

### Amplify Framework Version

1.17.0

### Amplify Categories

Auth

### Dependency manager

Cocoapods

### Swift version

4.2

### CLI version

Not applicable

### Xcode version

12.5

### Is this a regression?

Yes

### Regression additional context

Previous version details are mentioned in this bug: https://github.com/aws-amplify/amplify-ios/issues/1360

### Device

iPhone7+

### iOS Version

14.5

### Specific to simulators

No both Simulator and Devices are experiencing it.

**Comment by lawmicha** [comment]
Hi @shahan-ayyub, thanks for the details and sample app. Taking a look at the code, I can see references to both Amplify.Auth and Amplify.Storage. I'll provision both and run the app to better understand what the issue is. 

The code from the sample app
```swift
func signin(_ user: User) {
        //remove this get url and no issues will occur
        startDownloadingUrls(for: user)
        
        Amplify.Auth.signOut { result in
            switch result {
            case .success:
                Amplify.Auth.signIn(username: user.email) { result in
                    do {
                        let result = try result.get()
                        switch result.nextStep {
                        case .done:
                            self.updateMessage("Signin Successfull")
                        case .confirmSignInWithCustomChallenge:
                            Amplify.Auth.confirmSignIn(challengeResponse: "pass_\(user.password)") { result in
                                do {
                                    let result = try result.get()
                                    switch result.nextStep {
                                    case .done:
                                        self.updateMessage("Signin Confirm Successfull")

                                    default:
                                        self.updateMessage("Unknown next step while confirming sign in \(result.nextStep)")
                                    }
                                } catch {
                                    self.updateMessage("Error while confirming sign in \(error.localizedDescription)")
                                }
                            }
                        default:
                            self.updateMessage("Unknown next step while sign in \(result.nextStep)")
                        }
                    } catch {
                        self.updateMessage("Error while sign in \(error.localizedDescription)")
                    }
                }
            case .failure(let error):
                self.updateMessage("Error while signing out before sign in \(error.localizedDescription)")
            }
        }
    }
```

From reviewing the code, my Cognito auth flow is provisioned with a more simple flow, where `confirmSignInWithCustomChallenge` is not required, so after sign in, it will land in the `done` case.  

Scenarios I tried
1. sign in with user1 -> success
2. Sign in with user 2 -> failed. sign out. Sign in with user 2 -> success
3. Sign out, sign in with user 1, fetchAuthSession and print IdentityId -> success (identityId is of user1)
3. Sign out, sign in with user 2, fetchAuthSession and print IdentityId -> success (identityId is of user2)
4. download user1.jpeg with user2 credentials (since step 3 signs in with user 2), and in parallel sign out, sign in user 1 ->  the sign in is successful, however the storage call fails with 
```
Unexpected error occurred with message: Could not fetch AWS Cognito credentials, but there was no error reported back from
AWSMobileClient.getAWSCredentials call.
```

Downloading from S3 requires credentials, which can either be done by signing the request with user1 or user2. The request to download from S3 is done in parallel with the sign out call, which means that the session is invalidated while trying to sign in with Cognito to create another session. 

5. downloading user1.jpeg with user2 credentials (fails), and in parallel sign out, sign in with user 1, download user1.jpeg -> sign in was successful and s3 download completed successfully


Now looping back to the auth flow containing the `confirmSignInWithCustomChallenge`. I didn't actually update my auth flow to incorporate this, but just considering what it means in the logic that you have described. If we were to download a file using the credentials of the previous signed in user, and in parallel sign out, and then sign in with another user. The `confirmSignInWithCustomChallenge` would further increase the time window in which no user is authenticated at all, thus the S3 call will fail since it cannot successully sign the request.

**Comment by lawmicha** [comment]
I didn't quite understand this from the repro steps
```
See console logs in POC. You will notice that first "User 1" will be returned then "User 2" will be returned.
```
Are you referring to what's returned as the downloaded image or the user here? We can perform a fetchAuthSession and check the identityId of the current user to see who is signed in. 

The scenario I tried
1. Sign in with user 1 -> success (prints out identityId of user 1). Sign in with user 2, and while it's signing in, perform another sign in with user 2 -> one failure and one success (second sign in with user 2 will cause the first sign in to fail, the second sign in of user 2 is successful prints out identityId of user2 twice)
```
// first sign in with user 2 that fails
failed AuthError: Unexpected error occurred with message: Could not get end user to sign in.
```

If you are looking at what was successfully downloaded while running parallel tasks, then there may be a scenario that the download for user1's image is successful because user2 is successfully signed in and signing of the S3 request was done after user2 signed in.
`Amplify.Storage.getURL` downloads from the `public` folder, based on the access control you set, it can be read from all users. See file access levels here for more details https://docs.amplify.aws/lib/storage/configureaccess/q/platform/ios/

**Comment by lawmicha** [comment]
In the sample i didn't see any code that uses the **fetchAuthSession** API. You can use this to check user is signed in by matching it aganinst the user sub or identityId https://docs.amplify.aws/lib/auth/access_credentials/q/platform/ios/

let us know if i've missed something here or if you have any questions
